### PR TITLE
Added header-enricher

### PR DESCRIPTION
**Pull request by dturanski**

### CODE REVIEW

**Comment by garyrussell** [comment]
LGTM; except the missing pom and a couple of files need ending newlines.

File: header-enricher/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+## Building with Maven
+
+	$ mvn package
+
+The project's [pom](pom.xml) declares `spring-xd-module-parent` as its parent. This adds the dependencies needed to test the module and also configures the [Spring Boot Maven Plugin](http://docs.spring.io/spring-boot/docs/current/reference/html/build-tool-plugins-maven-plugin.html) to package the module as an uber-jar, packaging any dependencies that are not already provided by the Spring XD container. See the [Modules](http://docs.spring.io/spring-xd/docs/current/reference/html/#modules) section in the Spring XD Reference for a more detailed explanation of module class loading.
+
+## Installing the Custom Module
+
+The uber-jar will be in `[project-build-dir]/load-generator-source-feed-1.0.0.BUILD-SNAPSHOT.jar`. To install and register the module to your Spring XD distribution, use the `module upload` Spring XD shell command. Start Spring XD and the shell:
```
Comment (artembilan) [inline header-enricher/README.md]: Copy/paste typo?
Is `header-enricher-1.0.0.BUILD-SNAPSHOT.jar` correct value ?

File: header-enricher/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		
+**foo=123.45** !!
+
+    2015-05-24T11:45:16-0400 1.2.0.SNAP INFO pool-23-thread-4 sink.t3 - {requestMethod=POST, foo=123.45, User-Agent=Java/1.8.0_25, Host=localhost:9000, id=87d229b9-a434-31c2-eb35-b42ca3f8352a, Content-Length=19, contentType=text/plain;Charset=UTF-8, requestPath=/, timestamp=1432482316080}
+
+### Literal Strings with Embedded Spaces
+
+SpEL expects literals to be enclosed in single quotes. This is straight forward when the literal does not contain embedded spaces as in the `Multple Headers` example above. Embedded spaces works by encoding the Unicode escape sequence. The leading '\' must itself be escaped:
+
+    xd:>stream create t4 --definition "http | header-enricher --headers={\"foo\":\"'this\\u0020is\\u0020a\\u0020literal\\u0020string'\"} | log --expression=headers" --deploy
```
Comment (artembilan) [inline header-enricher/README.md]: This is very unfriendly :cry: ...

From other side: is it really SpEL issue? Maybe XD CLI parser issue ?

File: header-enricher/pom.xml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+		<artifactId>spring-xd-module-parent</artifactId>
+		<version>1.2.0.BUILD-SNAPSHOT</version>
+		<relativePath></relativePath>
+	</parent>
+
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+		<gmavenVersion>1.4</gmavenVersion>
+		<gmavenProviderSelection>2.0</gmavenProviderSelection>
+		<groovyVersion>2.3.6</groovyVersion>
```
Comment (artembilan) [inline header-enricher/pom.xml]: Can't we use the last `2.4.3` Groovy ?

File: header-enricher/src/test/java/HeaderEnricherIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+import org.springframework.messaging.Message;
+import org.springframework.xd.dirt.server.singlenode.SingleNodeApplication;
+import org.springframework.xd.dirt.test.SingleNodeIntegrationTestSupport;
+import org.springframework.xd.dirt.test.SingletonModuleRegistry;
+import org.springframework.xd.dirt.test.process.SingleNodeProcessingChain;
+import org.springframework.xd.module.ModuleType;
+import org.springframework.xd.test.RandomConfigurationSupport;
+import static org.junit.Assert.*;
+import static org.junit.Assert.assertEquals;
+import static org.springframework.xd.dirt.test.process.SingleNodeProcessingChainSupport.*;
```
Comment (artembilan) [inline header-enricher/src/test/java/HeaderEnricherIntegrationTest.java]: Are we OK with asterisk imports in XD ?

File: header-enricher/src/test/java/HeaderEnricherIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+		Message<?> message = chain.receive(RECEIVE_TIMEOUT);
+		assertEquals("HELLO,WORLD!", message.getHeaders().get("foo"));
+		assertEquals("ello", message.getHeaders().get("bar"));
+
+		chain.destroy();
+
+	}
+			
+
+	String headerValues = "{\"foo\":\"(payload+', world!').toUpperCase()\",\"bar\":\"payload.substring(1)\"}";
```
Comment (artembilan) [inline header-enricher/src/test/java/HeaderEnricherIntegrationTest.java]: What is the purpose of this variable ?

**Comment by artembilan** [comment]
That's all. 
Should I do my Cassandra sink similar way in this project as well ?

**Comment by garyrussell** [comment]
Yes; it's too late to hit the main project - will probably move in 1.3

File: header-enricher/README.md (hunk 2 above)
Comment (dturanski) [inline header-enricher/README.md]: It's an XD parser Issue actually. It doesn't like the spaces. This is a known issue.

File: header-enricher/src/test/java/HeaderEnricherIntegrationTest.java (hunk 4 above)
Comment (dturanski) [inline header-enricher/src/test/java/HeaderEnricherIntegrationTest.java]: generally, no

File: header-enricher/README.md (hunk 2 above)
Comment (ericbottard) [inline header-enricher/README.md]: you should try with some quote (either single or double, escaping accordingly) just after `--headers=`

Comment (dturanski) [inline header-enricher/README.md]: can you provide the literal string?

**Comment by ericbottard** [comment]
Could not build b/c of the following mvn error:
```
[ERROR] Failed to execute goal on project header-enricher: Could not resolve dependencies for project org.springframework.xd.modules:header-enricher:jar:1.0.0.BUILD-SNAPSHOT: Cannot access jcenter (http://jcenter.bintray.com) in offline mode and the artifact org.eclipse.paho:org.eclipse.paho.client.mqttv3:jar:1.0.2 has not been downloaded from it before. -> [Help 1]
```

**Comment by ericbottard** [comment]
OK, successfully tested with
```
stream create t4 --definition "http | header-enricher --headers='{\"foo\":\"''this is a literal string''\"}' | log --expression=headers" --deploy
```

you decide whether it's better or not :)

I'd like to add triple quoting to spring shell, which may help a bit with these situations and remove the need for the backslashes here

**Comment by dturanski** [comment]
Cool. That looks a lot better, I'll update the read me.

**Comment by ericbottard** [comment]
Rebased and merged as 62ee3ef436de7c2c3bae94666eb2f5d42ed6b43b

**Pull request closed without merging**